### PR TITLE
perf(lerobot_local): cache loaded model across policy instances

### PR DESCRIPTION
**Pull request by cagataycali**
## Problem

`LerobotLocalPolicy` reloads the full model on every instantiation. When an eval/rollout driver builds a fresh policy per episode (the natural `create_policy("lerobot_local", ...)`-in-a-loop pattern), each call re-reads the checkpoint weights from disk and re-uploads them to the device. For a large VLA this dominates wall time: `allenai/MolmoAct2-SO100_101` ships 1295 weight files and takes ~100-130s per load, so a 20-episode loop spends ~40 min purely re-loading the same weights.

## Change

Add a process-level cache of the built underlying model in `lerobot_local`:

- Keyed by `(pretrained_name_or_path, policy_type, device)` for the generic LeRobot path, and by the checkpoint + device + normalisation/processor knobs (`norm_tag`, `inference_action_mode`, `image_keys`, embodiment, dims) for the MolmoAct2 transformers-native path.
- Later instances with the same key reuse the resident module and skip the weight load. The resolved `policy_type` is cached alongside the module so a hit needs no extra hub round-trip.
- `molmoact2.build_policy` gains a `prebuilt_policy` parameter: on a cache hit the cheap config + pre/post processors are still rebuilt fresh (so per-instance processor state is never shared), and only the expensive weight construction is skipped.
- New `cache_model: bool = True` constructor flag and `clear_model_cache()` public helper (exported from `strands_robots.policies.lerobot_local`).

### Sharing contract

The cached object is the same live module shared by every instance with the same key. LeRobot policies hold per-episode mutable state (action queue, temporal-ensemble buffers) that `Policy.reset()` clears between episodes, so **sequential** reuse - including `Simulation.eval_policy(n_episodes=N)` and per-rollout drivers - is safe. For the rare case of driving two rollouts of the same checkpoint+device **concurrently**, pass `cache_model=False` for a private load. `clear_model_cache()` evicts cached models and releases the held GPU/CPU memory (best-effort `torch.cuda.empty_cache()`).

Note: for multi-episode evaluation, `Simulation.eval_policy(..., n_episodes=N)` and `run_policy(..., policy_object=loaded)` already load the policy once and reuse it; this cache additionally protects callers that re-`create_policy` per call.

## Measured effect

Real load on a CUDA host with `lerobot/diffusion_pusht` (generic path):

```
1st create_policy (cold load):  6.24s
2nd create_policy (cache hit):  0.222s
3rd create_policy (cache hit):  0.230s
same underlying module shared:  True
speedup cold/cached:            28x
clear_model_cache() -> reload:  fresh module rebuilt
cache_model=False:              private module (not shared)
```

For a MolmoAct2-scale 130s cold load, a 20-episode re-instantiation loop drops from ~43 min to ~2.4 min.

## Tests

`tests/policies/lerobot_local/test_model_cache.py` pins, for both the generic and MolmoAct2 load paths: heavy load happens once and the module is shared across instances; `cache_model=False` forces a private load; `clear_model_cache()` restores a cold load; distinct device keys do not collide. A `conftest.py` autouse fixture clears the module-level cache around every test in the package so the existing `"test/model"`-placeholder tests stay hermetic. All 295 `lerobot_local` tests pass; `ruff`/`ruff format`/`mypy` clean.

## Changelog

**Round 2** - Address CodeQL 'empty except' finding in `clear_model_cache()`: the swallowed `RuntimeError`/`AssertionError` on the best-effort `torch.cuda.empty_cache()` is intentional (entries are already evicted; a release hiccup must not fail the eviction). Added an explanatory comment documenting that intent. No behavior change.

### CODE REVIEW

**Review by yinsong1986 — DISMISSED** [review]
## Summary

This PR adds a process-level cache of the built underlying model to `LerobotLocalPolicy` so that re-instantiating a policy for the same checkpoint (the natural `create_policy("lerobot_local", ...)`-per-episode pattern) reuses the resident `nn.Module` instead of re-reading weights from disk and re-uploading to the device. The cache is keyed on `(namespace, path, device, ...)` for both the generic LeRobot path and the MolmoAct2 transformers-native path; `molmoact2.build_policy` gains a `prebuilt_policy` parameter that reuses the resident weights while still rebuilding the cheap config + pre/post processors fresh so per-instance processor state is never shared. A `cache_model=True` constructor flag and a `clear_model_cache()` helper (exported from the package) round it out.

## What's good

- The concurrency contract is documented explicitly (AGENTS.md > #85 "Document the concurrency contract"): the cache dict is lock-protected, sequential reuse is shown safe via `Policy.reset()`, and the rare concurrent case has a clear `cache_model=False` opt-out.
- The cache is a transparent no-op when there is no checkpoint path, so from-scratch policies are unaffected.
- `device` is part of the cache key and the distinct-device test pins that keys don't collide; the MolmoAct2 key folds in the normalisation/processor knobs that shape the pipeline.
- Solid regression coverage (generic + MolmoAct2 paths, opt-out, eviction, device isolation) plus an autouse `conftest.py` fixture that keeps the existing `"test/model"` placeholder tests hermetic.
- Scope is disciplined: no emojis or host paths in the diff, docs updated in the same PR.

## Verification suggestions

Standard CI (`hatch run test` for `tests/policies/lerobot_local/`) covers the cache behaviour. For a real-hardware spot-check, the PR's reported `lerobot/diffusion_pusht` cold-vs-cached timing (cold ~6.2s, cache hit ~0.22s, same module shared) is reproducible with two back-to-back `create_policy("lerobot_local", ...)` calls on a CUDA host followed by an identity check on `policy._policy`.

**Review by yinsong1986 — APPROVED** [review]
## Summary

This PR adds a process-level cache of the built underlying model to `LerobotLocalPolicy` so that re-instantiating a policy for the same checkpoint+device (the natural `create_policy("lerobot_local", ...)`-per-episode pattern) reuses the resident `nn.Module` instead of re-reading weights from disk and re-uploading to the device. The cache is keyed by `(namespace, path, device, ...)` with `"generic"` storing `(module, resolved_policy_type)` and `"molmoact2"` storing the bare module (rebuilding cheap config + pre/post processors fresh via the new `prebuilt_policy` parameter so per-instance processor state is never shared). A `cache_model=True` constructor flag and an exported `clear_model_cache()` helper round it out.

## What's good

- The shared-mutable-state risk is the right thing to worry about and the PR handles it deliberately: the module cache dict is guarded by `_MODEL_CACHE_LOCK`, the sequential-reuse-is-safe contract (`Policy.reset()` clears per-episode action-queue/temporal-ensemble state) is documented inline and in the docs, and a `cache_model=False` opt-out is provided for the concurrent case.
- Cache keys are scoped per load path and include `requested_device` (distinct device keys verified not to collide) and, for MolmoAct2, the processor-shaping knobs (`norm_tag`, `inference_action_mode`, `image_keys`, embodiment, dims) so a cached module can never be paired with a mismatched processor pipeline.
- The device-resolution/move block still runs on a cache hit, and `_cache_get`/`_cache_put` short-circuit on a `None` key so the cache is a transparent no-op for parameterless / caching-disabled policies.
- The CodeQL empty-except finding from the prior security-bot review is addressed with an explanatory comment on the best-effort `torch.cuda.empty_cache()` cleanup.
- Good test discipline: both load paths pinned (built-once + shared, `cache_model=False` private load, `clear_model_cache()` restores cold load, distinct-device no-collision), plus an autouse `conftest.py` fixture that clears the module-level cache around every test to keep the existing `"test/model"` placeholder tests hermetic.

No blocking concerns found.
